src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class NullableWalker
+    {
+        internal sealed class SnapshotManager
+        {
+            /// <summary>
+            /// The int key corresponds to <see cref="Snapshot.GlobalStateIndex"/>.
+            /// </summary>
+            private readonly ImmutableArray<SharedWalkerState> _walkerGlobalStates;
+            private readonly ImmutableDictionary<int, Snapshot> _incrementalSnapshots;
+
+            private SnapshotManager(ImmutableArray<SharedWalkerState> walkerGlobalStates, ImmutableDictionary<int, Snapshot> incrementalSnapshots)
+            {
+                _walkerGlobalStates = walkerGlobalStates;
+                _incrementalSnapshots = incrementalSnapshots;
+            }
+
+            internal NullableWalker RestoreWalkerToAnalyzeNewNode(
+                int position,
+                BoundNode nodeToAnalyze,
+                CSharpCompilation compilation,
+                Binder binder,
+                ImmutableDictionary<BoundExpression, (NullabilityInfo, TypeSymbol)>.Builder analyzedNullabilityMap)
+            {
+                // TODO: Use a more efficient storage and lookup mechanism like an AVL tree here.
+                // https://github.com/dotnet/roslyn/issues/35037
+                Snapshot incrementalSnapshot = default;
+                bool foundSnapshot = false;
+                foreach (var (currentPosition, currentSnapshot) in _incrementalSnapshots)
+                {
+                    if (currentPosition <= position)
```

Comment:
>if (currentPosition <= position) [](start = 20, length = 32)

Are we iterating sequence in any particular order? If not, then it looks like we likely to grab a snapshot in a non-deterministic way.

Reply:
Will switch to `ImmutableSortedDictionary` for now. We do need to revamp this search code entirely, but that will at least get us in a consistent state.

---
In reply to: [289474582](https://github.com/dotnet/roslyn/pull/35850#discussion_r289474582) [](ancestors = 289474582)

Reply:
It looks like this implementation is always going to pick a snapshot at the beginning of the method body.

---
In reply to: [289501930](https://github.com/dotnet/roslyn/pull/35850#discussion_r289501930) [](ancestors = 289501930,289474582)